tests/apollo/test_skvbc_byzantine_primary_preexecution.py

```diff
@@ -0,0 +1,396 @@
+from os import environ
+
+import trio
+
+from util import skvbc as kvbc
+from util.bft import with_trio, with_bft_network, KEY_FILE_PREFIX, with_constant_load
+from util.skvbc_history_tracker import verify_linearizability
+import util.eliot_logging as log
+
+SKVBC_INIT_GRACE_TIME = 2
+BATCH_SIZE = 4
+NUM_OF_SEQ_WRITES = 25
+NUM_OF_PARALLEL_WRITES = 100
+MAX_CONCURRENCY = 10
+LONG_REQ_TIMEOUT_MILLI = 15000
+ASSERT_TIMEOUT_SEC = 90
+
+def start_replica_cmd(builddir, replica_id, view_change_timeout_milli="10000"):
+    """
+    Return a command that starts an skvbc replica when passed to
+    subprocess.Popen.
+    The replica is started with a short view change timeout.
+    Note each arguments is an element in a list.
+    """
+
+    status_timer_milli = "500"
+
+    path = os.path.join(builddir, "tests", "simpleKVBC", "TesterReplica", "skvbc_replica")
+    cmd = [path,
+           "-k", KEY_FILE_PREFIX,
+           "-i", str(replica_id),
+           "-s", status_timer_milli,
+           "-v", view_change_timeout_milli,
+           "-x"
+           ]
+    if(replica_id == 0) :
+        cmd.extend(["-g", "MangledPreProcessResultMsgStrategy"])
+
+    return cmd
+
+def start_replica_cmd_assymetric_communication(builddir, replica_id, view_change_timeout_milli="10000"):
+    """
+    Return a command that starts an skvbc replica when passed to
+    subprocess.Popen.
+    The replica is started with a short view change timeout.
+    Note each arguments is an element in a list.
+    """
+
+    status_timer_milli = "500"
+
+    path = os.path.join(builddir, "tests", "simpleKVBC", "TesterReplica", "skvbc_replica")
+    cmd = [path,
+           "-k", KEY_FILE_PREFIX,
+           "-i", str(replica_id),
+           "-s", status_timer_milli,
+           "-v", view_change_timeout_milli,
+           "-x"
+           ]
+    if(replica_id == 0) :
+        cmd.extend(["-d", "-g", "MangledPreProcessResultMsgStrategy"])
+
+    return cmd
+
+
+class SkvbcPrimaryByzantinePreExecutionTest(unittest.TestCase):
+
+    __test__ = False  # so that PyTest ignores this test scenario
+
+    async def check_viewchange_noexcept(self, bft_network, initial_primary, viewchange_timeout_secs=5):
+        try:
+            with trio.move_on_after(seconds=4*viewchange_timeout_secs):
+                await bft_network.wait_for_view(
+                    replica_id=random.choice(
+                        bft_network.all_replicas(without={initial_primary})),
+                    expected=lambda v: v != initial_primary,
+                    err_msg="Make sure view change has occurred."
+                )
+                new_primary = await bft_network.get_current_primary()
+                await bft_network.init_preexec_count(new_primary)
+                await trio.sleep(seconds=viewchange_timeout_secs)
+                return new_primary
+        except trio.TooSlowError:
+            return self.check_viewchange_noexcept(bft_network, initial_primary, viewchange_timeout_secs)
```

Comment:
are you using recursion here?

If so, note that `self.check_viewchange_noexcept()` is async, so you need to call it with `await`. Otherwise it won't be executed ...

Reply:
Thanks @teoparvanov for this comment, good catch.